lib/internal/modules/esm/import_map.js

```diff
@@ -0,0 +1,196 @@
+'use strict';
+const { isURL, URL } = require('internal/url');
+const {
+  ObjectEntries,
+  ObjectKeys,
+  SafeMap,
+  ArrayIsArray,
+  StringPrototypeStartsWith,
+  StringPrototypeEndsWith,
+  StringPrototypeSlice,
+  ArrayPrototypeReverse,
+  ArrayPrototypeSort,
+} = primordials;
+const { codes: { ERR_INVALID_IMPORT_MAP } } = require('internal/errors');
+const { shouldBeTreatedAsRelativeOrAbsolutePath } = require('internal/modules/helpers');
+
+class ImportMap {
+  #baseURL;
+  #imports = new SafeMap();
+  #scopes = new SafeMap();
+  #specifiers = new SafeMap()
+
+  /**
+   * Process a raw import map object
+   * @param {object} raw The plain import map object read from JSON file
+   * @param {URL} baseURL The url to resolve relative to
+   */
+  constructor(raw, baseURL) {
+    this.#baseURL = baseURL;
+    this.process(raw);
+  }
+
+  // These are convenience methods mostly for tests
+  get baseURL() {
+    return this.#baseURL;
+  }
+
+  get imports() {
+    return this.#imports;
+  }
+
+  get scopes() {
+    return this.#scopes;
+  }
+
+  /**
+   * Cache for mapped specifiers
+   * @param {string | URL} originalSpecifier The original specifier being mapped
+   */
+  #getMappedSpecifier(originalSpecifier) {
+    let mappedSpecifier = this.#specifiers.get(originalSpecifier);
+
+    // Specifiers are processed and cached in this.#specifiers
+    if (!mappedSpecifier) {
+      // Try processing as a url, fall back for bare specifiers
+      try {
+        if (shouldBeTreatedAsRelativeOrAbsolutePath(originalSpecifier)) {
+          mappedSpecifier = new URL(originalSpecifier, this.#baseURL);
+        } else {
+          mappedSpecifier = new URL(originalSpecifier);
+        }
+      } catch {
+        // Ignore exception
+        mappedSpecifier = originalSpecifier;
+      }
+      this.#specifiers.set(originalSpecifier, mappedSpecifier);
+    }
+    return mappedSpecifier;
+  }
+
+  /**
+   * Resolve the module according to the import map.
+   * @param {string | URL} specifier The specified URL of the module to be resolved.
+   * @param {string | URL} [parentURL] The URL path of the module's parent.
+   * @returns {string | URL} The resolved module specifier
+   */
+  resolve(specifier, parentURL = this.#baseURL) {
+    // When using the customized loader the parent
+    // will be a string (for transferring to the worker)
+    // so just handle that here
+    if (!isURL(parentURL)) {
+      parentURL = new URL(parentURL);
+    }
+
+    // Process scopes
+    for (const { 0: prefix, 1: mapping } of this.#scopes) {
+      const _mappedSpecifier = mapping.get(specifier);
+      if (StringPrototypeStartsWith(parentURL.pathname, prefix.pathname) && _mappedSpecifier) {
```

Comment:
This seeme like not following spec.
There should be check for trailing slash, mb I'm missing smth
https://html.spec.whatwg.org/multipage/webappapis.html#the-resolution-algorithm (10.1)
https://github.com/WICG/import-maps/blob/abc4c6b24e0cc9a764091be916c5057e83c30c23/reference-implementation/lib/resolver.js#L47